File: Python/CreationalDesignPattern/ProxyPatternDemo_cache.py
from abc import ABC, abstractmethod
from typing import Dict, Optional

# fun non
# entitieis
#
# relationsji
#
# class Notification
#     - priority
#     - message


class IRateLimitinStrategy(ABC):
    pass


class Notification:
    def __init__(self):
        self.priority: Optional[str] = None
        self.message: Optional[str] = None


class Database(ABC):
    @abstractmethod
    def fetch_data(self, query: str) -> str:
        pass


class RealDatabase(Database):
    def fetch_data(self, query: str) -> str:
        print("Fetching confidential employee salary data...")
        return f"data for {query}"


class DatabaseProxy(Database):
    def __init__(self, role: Optional[str]):
        self.role = role
        self.database: Optional[RealDatabase] = None
        self.cache: Dict[str, str] = {}

    def fetch_data(self, query: str) -> str:
        if query in self.cache:
            return self.cache[query]

        if self.database is None:
            self.database = RealDatabase()

        value = self.database.fetch_data(query)
        self.cache[query] = value
        return value


def main():
    user = DatabaseProxy(None)
    user.fetch_data("dsadsadasd")


if __name__ == "__main__":
    main()

File: C++/CreationalDesignPattern/ProxyPatternDemo_cache.cpp
#include <iostream>
#include <memory>
#include <optional>
#include <string>
#include <unordered_map>

// fun non
// entitieis
//
// relationsji
//
// class Notification
//     - priority
//     - message

struct IRateLimitinStrategy {
    virtual ~IRateLimitinStrategy() = default;
};

struct Notification {
    std::string priority;
    std::string message;
};

struct Database {
    virtual std::string fetchData(const std::string& query) = 0;
    virtual ~Database() = default;
};

struct RealDatabase : Database {
    std::string fetchData(const std::string& query) override {
        std::cout << "Fetching confidential employee salary data..." << std::endl;
        return "data for " + query;
    }
};

class DatabaseProxy : public Database {
    std::optional<std::string> role;
    std::unique_ptr<RealDatabase> database;
    std::unordered_map<std::string, std::string> cache;

public:
    explicit DatabaseProxy(std::optional<std::string> role) : role(std::move(role)) {}

    std::string fetchData(const std::string& query) override {
        auto it = cache.find(query);
        if (it != cache.end()) {
            return it->second;
        }

        if (!database) {
            database = std::make_unique<RealDatabase>();
        }

        std::string value = database->fetchData(query);
        cache[query] = value;
        return value;
    }
};

int main() {
    DatabaseProxy user(std::nullopt);
    user.fetchData("dsadsadasd");

    return 0;
}
